docs/concepts/README.md

```diff
@@ -0,0 +1,217 @@
+# Duck Typing
+
+Knative uses duck typing to keep various components loosely coupled, but
+[what is duck typing](https://en.wikipedia.org/wiki/Duck_typing)? It is
+reasoning about a resource's control plane shape and behaviors based on a common
+definition of that contact. If a resource has the same fields in the same schema
+locations as the contract specifies, and the control/data plane behaviors as the
+contract specifies, Knative can use that resource as if it is the generic duck
+type with little specific knowledge about the resource type. Some resources may
+choose to opt-in or be opt'ed-in to multiple duck types.
+
+<!-- TODO: point to Discovery ClusterDuckType documentation. -->
+
+The most basic usage of duck typing in Knative is our usage of object refs in
+resource specs to "point" to another resource. The contract of the object
+holding the ref will prescribe the expected duck type of the "pointee".
+
+For example, take this example Knative resource `pointer` that is pointing to a
+`Dog` named `pointee`:
+
+```yaml
+apiVersion: sample.knative.dev/v1
+kind: Example
+metadata:
+  name: pointer
+spec:
+  size:
+    apiVersion: extension.example.com/v1
+    kind: Dog
+    name: pointee
+```
+
+Assume the expected shape of a "Sizable" duck type is that in the status, we
+expect the following schema shape:
+
+```yaml
+<standard metadata>
+<spec ignored for Sizable>
+status:
+  height: <in inches>
+  weight: <in pounds>
+```
```

Comment:
Since we don't talk about schemas anywhere else really until now, it might be good to explain what this actually is in terms of Knative resources? If the YAML example is basically an `Example` resource the same as the one above, this isn't obvious from the explanation, because it's not obvious that a schema and a resource are the same thing. I'd stick with just saying "resource" to avoid confusion if that's the case. I also think we should never abbreviate examples with things like `<standard metadata` because we're assuming a reader is more familiar with this than they might actually be, and again it's not clear if its the same thing as the resource metadata above.

Reply:
Reading through this bit and the Human example bit I get what you're saying, but I think the way it's written is sort of confusing. I don't have time really to rewrite/edit it all for now, so I'm happy enough to merge as is when you're ready, but let's maybe put it on our to do list to revisit how we're explaining this later to see if it can be clearer 🙂

Reply:
Sure! I was just rolling with one concept as a start